File: 2901.longest-unequal-adjacent-groups-subsequence-ii.cs
/*
 * @lc app=leetcode id=2901 lang=csharp
 *
 * [2901] Longest Unequal Adjacent Groups Subsequence II
 */

// @lc code=start
public class Solution
{
    public IList<string> GetWordsInLongestSubsequence(string[] words, int[] groups)
    {
        //use brute dp
        // 3 Requirement
        //兩個word帶的groups數值，必須各不相同
        //兩個word的字元長度，必須相同
        //兩個word的漢明距離，必須為1  ex:bab跟cab，他們只有第一個字元不同剩下都為ab，所以漢明距離為1
        int n = words.Length;
        int[] dp = new int[n]; //dp[i] = length of longest valid subsequence ending at i
        int[] prev = new int[n]; //prev[i] index of previous word in the sequence
        Array.Fill(dp, 1);   //assigns value 1 to each element of the dp array
        Array.Fill(prev, -1);

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (groups[i] != groups[j] && //符合group值不同
                    words[i].Length == words[j].Length && //符合字串長度相同                    
                    HammingDistance(words[i], words[j]) == 1) //符合漢明距離為1
                {
                    if (dp[j] + 1 > dp[i])
                    {
                        dp[i] = dp[j] + 1;
                        prev[i] = j;
                    }
                }
            }
        }

        //找最大dp value
        int maxLength = 0;
        int endIndex = 0;
        for (int i = 0; i < n; i++)
        {
            if (dp[i] > maxLength)
            {
                maxLength = dp[i];
                endIndex = i;
            }
        }

        //重新把字串塞回序列
        List<string> res = new List<string>();
        while (endIndex != -1)
        {
            res.Add(words[endIndex]);
            endIndex = prev[endIndex];
        }
        res.Reverse();        
        return res;
    }

    int HammingDistance(string word_I, string word_J) //compute the HammingDistance
    {
        int count = 0;
        for (int i = 0; i < word_I.Length; i++)
        {
            if (word_I[i] != word_J[i]) count++;
        }
        return count;
    }
}
// @lc code=end

File: PracticeFolder/2901.longest-unequal-adjacent-groups-subsequence-ii.cpp
/*
 * @lc app=leetcode id=2901 lang=cpp
 *
 * [2901] Longest Unequal Adjacent Groups Subsequence II
 */

// @lc code=start
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;
class Solution {
public:
    bool differByOneChar(string word1, string word2) {
        if (word1.length() != word2.length()) return false;
        int diffCount = 0;
        for (int i = 0; i < word1.length(); i++) 
            diffCount += word1[i] != word2[i];
        return diffCount == 1;
    }
    
    vector<string> getWordsInLongestSubsequence(vector<string>& words, vector<int>& groups) {
        int n = groups.size();
        vector<int> dp(n, 1), parent(n, -1);
        int maxi = 0;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (groups[i] != groups[j] && 
                        differByOneChar(words[i], words[j]) && 
                            dp[i] < dp[j] + 1) {
                    dp[i] = dp[j] + 1;
                    parent[i] = j;
                }
            }
            maxi = max(maxi, dp[i]);
        }
        
        vector<string> result;
        for (int i = 0; i < n; i++) {
            if (maxi == dp[i]) {
                while (i != -1) {
                    result.push_back(words[i]);
                    i = parent[i];
                }
                break;
            }
        }
        reverse(result.begin(), result.end());
        return result;
    }
};

// @lc code=end

/*class Solution {
public:
    vector<string> getWordsInLongestSubsequence(vector<string>& words, vector<int>& groups) 
    {
        /*
        for (i = 0; i++; i < words.length;)
        {
            for
            vector<string> temp = words[i], words[j];

            if (groups[i] != groups[j]) //符合group值不同
            {
                if (words[i].length == words[j].length) //符合字串長度相同
                {
                    if (checkHammingDistance(words[i], words[j])) //符合漢明距離為1
                    {
                        return vector
                    }
                    
                }
            } 
        }
        
          
    }
};*/